src/corecel/cont/Bitset.hh

```diff
@@ -0,0 +1,395 @@
+        // sanitize the last word
+        detail::Sanitize<N % bits_per_word>::sanitize(this->last_word());
+        return *this;
+    }
+
+    //! Set the bit at position pos
+    CELER_CONSTEXPR_FUNCTION Bitset&
+    set(size_type pos, bool value = true) noexcept
+    {
+        if (value)
+        {
+            this->get_word(pos) |= Bitset::mask(pos);
+        }
+        else
+        {
+            this->get_word(pos) &= ~Bitset::mask(pos);
+        }
+
+        return *this;
+    }
+
+    //! Reset all bits
+    CELER_CONSTEXPR_FUNCTION Bitset& reset() noexcept
+    {
+        for (size_type i = 0; i < num_words; ++i)
+        {
+            words_[i] = static_cast<word_type>(0);
+        }
+
+        return *this;
+    }
+
+    //! Reset the bit at position pos
+    CELER_CONSTEXPR_FUNCTION Bitset& reset(size_type pos) noexcept
+    {
+        this->get_word(pos) &= ~Bitset::mask(pos);
+        return *this;
+    }
+
+    //! Flip all bits
+    CELER_CONSTEXPR_FUNCTION Bitset& flip() noexcept
+    {
+        for (size_type i = 0; i < num_words; ++i)
+        {
+            words_[i] = ~words_[i];
+        }
+        // sanitize the last word
+        detail::Sanitize<N % bits_per_word>::sanitize(last_word());
+        return *this;
+    }
+
+    //! Flip the bit at position pos
+    CELER_CONSTEXPR_FUNCTION Bitset& flip(size_type pos) noexcept
+    {
+        this->get_word(pos) ^= Bitset::mask(pos);
+        return *this;
+    }
+
+  private:
+    //! Find the word index for a given bit position
+    static CELER_CONSTEXPR_FUNCTION size_type which_word(size_type pos) noexcept
+    {
+        return pos / bits_per_word;
+    }
+
+    //! Find the bit index in a word
+    static CELER_CONSTEXPR_FUNCTION size_type which_bit(size_type pos) noexcept
+    {
+        return pos % bits_per_word;
+    }
+
+    //! Create a mask for a given bit index
+    static CELER_CONSTEXPR_FUNCTION word_type mask(size_type pos) noexcept
+    {
+        return static_cast<word_type>(1) << Bitset::which_bit(pos);
+    }
+
+    //! Get the word for a given bit position
+    CELER_CONSTEXPR_FUNCTION word_type get_word(size_type pos) const noexcept
+    {
+        return words_[Bitset::which_word(pos)];
+    }
+
+    //! Get the word for a given bit position
+    CELER_CONSTEXPR_FUNCTION word_type& get_word(size_type pos) noexcept
+    {
+        return words_[Bitset::which_word(pos)];
+    }
+
+    //! Get the last word of the bitset
+    CELER_CONSTEXPR_FUNCTION word_type& last_word() noexcept
+    {
+        return words_[num_words - 1];
+    }
+
+    //! Get the last word of the bitset
+    CELER_CONSTEXPR_FUNCTION word_type last_word() const noexcept
+    {
+        return words_[num_words - 1];
+    }
+
+    //! storage
+    word_type words_[num_words] = {};
+};
+
+/*!
+ * Reference to a single bit in the bitset.
+ * This is used to implement the mutable operator[].
+ */
+template<size_type N>
+class Bitset<N>::reference
+{
+    friend class Bitset;
```

Comment:
Is it necessary to add it as a friend? Since it's a nested class, I think it should have access to everything private.

Reply:
Actually not necessary for the current implementation